### PR TITLE
[DNM] cdc/kafka: fix kafka metrics collector

### LINKED ISSUE #4966: clean up all kafka sink metrics by maintain broker information
```
[2022/03/18 18:54:01.275 +00:00] [WARN] [metrics.go:279] ["clean up broker metrics failed"] [error="read tcp 10.233.127.131:38712->10.233.64.161:9092: i/o timeout"]
```

when cleaning up the Kafka metrics, broker information is required. But if Kafka brokers are unreachable, metrics may not be removed. This can be solved by always maintain broker information in the metrics collector.

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve?

Issue Number: close #4966

### What is changed and how it works?

when collecting and cleaning up metrics from sarama producer, it will fetch broker information.

If we cannot fetch brokers' information when collecting metrics, this always indicates that Kafka cluster is unreachable, use historical kafka cluster information.

When clean up sarama metrics, use historical brokers' information to do the work.

### Check List 

Tests 

 - No code

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
`None`.
```

### CODE REVIEW

File: cdc/sink/producer/kafka/kafka.go [hunk 1]
```diff
@@ -285,7 +288,13 @@ func (k *kafkaSaramaProducer) run(ctx context.Context) error {
 		case <-k.closeCh:
 			return nil
 		case <-ticker.C:
-			k.metricsMonitor.CollectMetrics()
+			// if error returned, it means the kafka broker clusters is unreachable,
+			// it's ok to return error. Most of the time, this only happen when no
+			// data sending, since if there is any data sending to the kafka,
+			// `ErrKafkaAsyncSendMessage` would have a higher chance to be triggered.
+			if err := k.metricsMonitor.CollectMetrics(); err != nil {
```
Comment (sdojjy) [inline cdc/sink/producer/kafka/kafka.go:295]: metrics collect failed should not stop the changefeed

Comment (3AceShowHand) [reply cdc/sink/producer/kafka/kafka.go:295]: yes, I will revise it twice.

File: cdc/sink/producer/kafka/metrics.go [hunk 2]
```diff
@@ -262,22 +280,42 @@ func (sm *saramaMetricsMonitor) collectBrokerMetrics() error {
...
+		defer func() {
+			ticker.Stop()
+			monitor.cleanup()
+		}()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				monitor.collectMetrics()
```
Comment (sdojjy) [inline cdc/sink/producer/kafka/metrics.go:309]: since collectMetrics will hang, print warning log if timeout? for example:  check the duration is gt 5s?

**Review by sdojjy — APPROVED** [review]
lgtm

File: errors.toml [hunk 3]
```diff
@@ -348,7 +348,7 @@ invalid admin job type: %d
 
 ["CDC:ErrInvalidChangefeedID"]
 error = '''
-bad changefeed id, please match the pattern "^[a-zA-Z0-9]+(\-[a-zA-Z0-9]+)*$, the length should no more than %d", eg, "simple-changefeed-task"
+bad changefeed id, please match the pattern "^[a-zA-Z0-9]+(\-[a-zA-Z0-9]+)*$", the length should no more than %d, eg, "simple-changefeed-task",
```
Comment (overvenus) [inline errors.toml:351]:
```suggestion
bad changefeed id, please match the pattern "^[a-zA-Z0-9]+(\-[a-zA-Z0-9]+)*$", the length should no more than %d, eg, "simple-changefeed-task"
```

Why is there a ','?

**Review by overvenus — APPROVED** [review]
Rest LGTM

File: cdc/sink/producer/kafka/metrics.go [hunk 4]
```diff
@@ -213,14 +217,28 @@ func getBrokerMetricName(prefix, brokerID string) string {
...
+			"use historical brokers to collect kafka broker level metrics",
+			zap.String("changefeed", sm.changefeedID),
+			zap.Any("role", sm.role),
+			zap.Duration("duration", time.Since(start)))
+		return
 	}
 
 	for _, b := range brokers {
-		brokerID := strconv.Itoa(int(b.ID()))
+		sm.brokers[b.ID()] = struct{}{}
```
Comment (zhaoxinyu) [inline cdc/sink/producer/kafka/metrics.go:233]: I think an int32 slice is more suitable here.

Comment (3AceShowHand) [reply cdc/sink/producer/kafka/metrics.go:233]: we maintain brokers by set, it's easier for insertion.

**Pull request merged**